nexus/reconfigurator/execution/src/lib.rs

```diff
@@ -123,143 +131,512 @@ pub async fn realize_blueprint_with_overrides(
+
+                StepSuccess::new(()).into()
+            },
+        )
+        .register();
+}
+
+fn register_decommission_expunged_disks_step<'a>(
+    registrar: &ComponentRegistrar<'_, 'a>,
+    opctx: &'a OpContext,
+    datastore: &'a DataStore,
+    deploy_disks_done: StepHandle<DeployDisksDone>,
+) {
+    // This depends on the "deploy_disks" call earlier -- disk expungement is a
+    // statement of policy, but we need to be assured that the Sled Agent has
+    // stopped using that disk before we can mark its state as decommissioned.
+    registrar
+        .new_step(
+            ExecutionStepId::Remove,
+            "Decommission expunged disks",
+            move |cx| async move {
+                let done = deploy_disks_done.into_value(cx.token()).await;
+                omicron_physical_disks::decommission_expunged_disks(
+                    &opctx, datastore, done,
+                )
+                .await
+                .map_err(merge_anyhow_list)?;
+
+                StepSuccess::new(()).into()
+            },
+        )
+        .register();
+}
+
+#[derive(Debug)]
+struct ReassignSagaOutput {
+    needs_saga_recovery: bool,
+    error: Option<anyhow::Error>,
+}
+
+fn register_reassign_sagas_step<'a>(
+    registrar: &ComponentRegistrar<'_, 'a>,
+    opctx: &'a OpContext,
+    datastore: &'a DataStore,
+    blueprint: &'a Blueprint,
+    nexus_id: Uuid,
+) -> StepHandle<ReassignSagaOutput> {
+    // For this and subsequent steps, we'll assume that any errors that we
+    // encounter do *not* require stopping execution.  We'll just accumulate
+    // them and return them all at the end.
     //
     // TODO We should probably do this with more of the errors above, too.
-    let mut errors = Vec::new();
+    registrar
+        .new_step(
+            ExecutionStepId::Ensure,
+            "Reassign sagas",
+            move |_cx| async move {
+                // For any expunged Nexus zones, re-assign in-progress sagas to
+                // some other Nexus.  If this fails for some reason, it doesn't
+                // affect anything else.
+                let sec_id = nexus_db_model::SecId(nexus_id);
+                let reassigned = sagas::reassign_sagas_from_expunged(
+                    &opctx, datastore, blueprint, sec_id,
+                )
+                .await
+                .context("failed to re-assign sagas");
+                match reassigned {
+                    Ok(needs_saga_recovery) => {
+                        let output = ReassignSagaOutput {
+                            needs_saga_recovery,
+                            error: None,
+                        };
+                        StepSuccess::new(output).into()
+                    }
+                    Err(error) => {
+                        // We treat errors as non-fatal here, but we still want
+                        // to log them.
+                        let message = error.to_string();
```

Comment:
Should this be
```suggestion
                        let message = format!("{error:#}");
```
to get the error chain? (Same question on other error -> StepWarning conversions)

Reply:
Ah that isn't necessary here because we return the full error chain at the end anyway. (Will add a comment explaining this)

Reply:
Done.